Created new NatNet SDK page (#276)
NUbots/NUbook: src/book/02-system/03-tools/06-natnet.mdx

2. Start the Motive software, it doesn't matter if motive is running locally or on another machine. Ensure that the [Data Streaming option](https://docs.optitrack.com/motive/data-streaming) in Motive is enabled. This will start sending packets to the Motive defined default multicast address on the local network.

3. Once you have the NUbots codebase running on your OS of choice and Motive running and Streaming packets to the network, NatNet can be run in one of two ways:

<details>
<summary>On the Robot</summary>

This is the default method for running NatNet. If you wish to output motive packets directly to file for debugging purposes, change the dump_packets option in NatNet.yaml to true.

1. Ensure that the NatNet role has been installed on the robot.

2. [SSH into the robot](guides/main/getting-started#running-on-a-real-robot).

3. `cd config`

4. `nano DataLogging.yaml`

5. You will need to add message.input.MotionCapture and set its value to true. Additionally, set to true any of the sub messages that you want to use such as ".RigidBody". If you want "output.CompressedImage" or "input.sensors" logged, you can set them to true as well.

> 6. To exit nano, press Ctrl+X, then 'y', then hit enter.

Reviewer: ```suggestion
6. To exit nano, press <kbd>Ctrl</kbd>+<kbd>X</kbd>, then <kbd>y</kbd>, then hit enter.
```

7. Run Natnet using `./b run natnet`

> 8. After your program has finished, check that a new nbs file has been created in "log/<your_program's_name>/". The filename should be the date and timestamp when the file was run.

Reviewer: ```suggestion
8. After your program has finished, check that a new nbs file has been created in `log/<your_program's_name>/`. The filename should be the date and timestamp when the file was run.
```

9. Exit out of the ssh of the robot.

> 10. We now need to copy the nbs file off the robot. Use "scp nubots@10.1.1.<robot_id#>:/home/nubots/log/<your_program_name>/<name_of_nbs_file> ./log/"

Reviewer: ```suggestion
10. We now need to copy the nbs file off the robot. Use `scp nubots@10.1.1.<robot_id#>:/home/nubots/log/<your_program_name>/<name_of_nbs_file> ./log/`
```

11. You now need to extract the information you want from the nbs file. We have a number of .nbs extraction tools and is recommended that you use `./b nbs extract_images <path>` for extracting images from the nbs file and `./b nbs json <path>` to extract the information into JSON format.

</details>

<details>
<summary>From your computer</summary>

This method may be useful to determine if Motive is outputting packets directly or debugging depacketization without the need of a robot.

1. Ensure that the NatNet role has been installed locally.

2. Ensure that the option `message.input.MotionCapture` in DataLogging.yaml is set to true.

3. Run Natnet using `./b run natnet`

4. After your program has finished, check that a new nbs file has been created in "log/<your_program's_name>/". The filename should be the date and timestamp when the file was run.

> 5. You now need to extract the information you want from the nbs file. We have a few nbs decoders that have already been written for specific purposes. If one of those does not suit your needs, you will need to either modify an existing nbs extraction file, or write your own. The location of these nbs files is "nubots/tools/nbs_tools/". For example purposes, let's assume you want to extract images and Motion Capture data. To run this program, make sure you are in the nubots home directory, then you would use: "./b nbs extract_images_MoCap ./log/<file_name> --output ./log/<your_extraction_folder>"

Reviewer: ```suggestion
5. You now need to extract the information you want from the nbs file. We have a few nbs decoders that have already been written for specific purposes. If one of those does not suit your needs, you will need to either modify an existing nbs extraction file, or write your own. The location of these nbs files is "nubots/tools/nbs_tools/". For example purposes, let's assume you want to extract images and Motion Capture data. To run this program, make sure you are in the nubots home directory, then you would use: `./b nbs extract_images_MoCap ./log/<file_name> --output ./log/<your_extraction_folder>`
```

</details>

> <Alert type='info'>
> NatNet may not always pick up motive packets immediately. This may be dependent of the amount of traffic on your local network or how it is configured. Best Practice to ensure the least flakeyness is to start streaming packets on motive significantly before you run NatNet and if NatNet still doesn't work, try stopping and starting NatNet a few times. If this still doesn't work, try restarting motive streaming and then run NatNet again.

Reviewer: In order to use code snippets inside alerts in nubook there needs to be a newline afterwards - I know there's none in this one but just for best practice
```suggestion
<Alert type='info'>

NatNet may not always pick up motive packets immediately. This may be dependent of the amount of traffic on your local network or how it is configured. Best Practice to ensure the least flakeyness is to start streaming packets on motive significantly before you run NatNet and if NatNet still doesn't work, try stopping and starting NatNet a few times. If this still doesn't work, try restarting motive streaming and then run NatNet again.
```

---
section: System
chapter: Tools
title: NatNet SDK

> description: A tool that provides positional infomation from motive mocap software.

Reviewer 2: ```suggestion
description: A tool that captures positional information from the Motive motion capture software.
```

slug: /system/tools/natnet_sdk
authors:

>   - Luke Haigh

Reviewer 2: ```suggestion
  - Luke Haigh (@Dereferenced-Null)
```

---

> NatNet is module that works in conjunction with the [OptiTrack Motive software](https://www.optitrack.com/software/motive/) and motion tracking camera system. The purpose of the NatNat system is to record positional information about the NUgus robots in relation to the playing field and the external envrionment so that it can be provided to our automatic anotation system which will use the positional information of the robots to automatically create segmentation masks of the environment. 

Reviewer 2: Make this description more generic - also make the distinction between NatNet the networking sdk from Motive and NatNet the module in the NUbots codebase. NatNet is it's own thing, we just have a module to capture the NatNet packets. As for making it more generic, don't mention the automatic annotator, just say in general that it can be used to get motion capture data, and for example could be used to get positional information from the environment (field, robot, etc).

Reviewer 2: Also watch for typos.

> ## Getting NatNet Running

Reviewer 2: ```suggestion
##  Running NatNet
```

> In order to get NatNet running you will require a licenced version of the Optitrack Motive software and at least 2 [Optive Motion Capture Cameras](https://www.optitrack.com/cameras/). NatNet currently supports motive version 3.0 and below. Below is a detailed guide on how to setup motive.

Reviewer 2: ```suggestion
In order to run NatNet you will require a licenced version of the Optitrack Motive software and at least 2 [Optive Motion Capture Cameras](https://www.optitrack.com/cameras/). NatNet currently supports motive version 3.0 and below. Below is a detailed guide on how to setup motive.
```
surely this will work regardless of the number of cameras? practically, sure you'll want two but it dooesn't stop it running.

<details>
<summary>Setting up Motive</summary>

1. Follow the [Optitrack guide for Installation and Activation](https://docs.optitrack.com/motive/installation-and-activation) if you don't already have Motive installed.

2. Open the Motive software. You will be presented with a range of options, but it is safest to choose "Perform Camera Calibration" and re-calibrate the cameras, in case one or two of them have been accidentally moved since the last calibration.

> 3. You should have 10 black boxes visible at the bottom of your screen representing the view of the 10 cameras mounted around the room (they will default to object mode which only displays objects whiter than a certain threshold, assuming them to be motion capture dots. If you don't have 10 camera boxes displaying, check the netgear ethernet switch is powered on. If the cameras are on, they will display a number on their LCD screen.

Reviewer 2: probably specify somewhere that this is for the lab. or instead of ten, just say 'the number of cameras you're using'.

> 4. Make sure there are no motion capture dots on anything on the soccer field.

Reviewer 2: ```suggestion
4. Make sure there are no motion capture dots anywhere on the soccer field.
```
or 'in view of the cameras' might be better

3. Once you have the NUbots codebase running on your OS of choice and Motive running and Streaming packets to the network, NatNet can be run in one of two ways:

<details>
<summary>On the Robot</summary>

This is the default method for running NatNet. If you wish to output motive packets directly to file for debugging purposes, change the dump_packets option in NatNet.yaml to true. 

1. Ensure that the NatNet role has been installed on the robot.

2. [SSH into the robot](guides/main/getting-started#running-on-a-real-robot).

3. `cd config`

4. `nano DataLogging.yaml`

5. Scroll down to message.input.MotionCapture and set to true. Additionally, set to true any of the sub messages that you want to use such as ".RigidBody". If you want "output.CompressedImage" or "input.sensors" logged, you can set them to true as well. Don't log "input.Image" as this will consume huge amounts of storage.

> 

Reviewer 2: have you been working off an old branch? very few messages are in the DataLogging yaml file, the user will need to add in `message.input.MotionCapture`. `message.input.Image` shouldn't be there either.

Author: I pretty much took this stuff straight from the original setup document that we used to set up mocap so I wasn't sure what had changed

6. To exit nano, press Ctrl+X, then 'y', then hit enter.

7. Run Natnet using `./b run natnet`

8. After your program has finished, check that a new nbs file has been created in "log/<your_program's_name>/". The filename should be the date and timestamp when the file was run.

9. Exit out of the ssh of the robot.

10. We now need to copy the nbs file off the robot. Use "scp nubots@10.1.1.<robot_id#>:/home/nubots/log/<your_program_name>/<name_of_nbs_file>    ./log/"

> 11. You now need to extract the information you want from the nbs file. We have a few nbs decoders that have already been written for specific purposes. If one of those does not suit your needs, you will need to either modify an existing nbs extraction file, or write your own. The location of these nbs files is "nubots/tools/nbs_tools/". For example purposes, let's assume you want to extract images and Motion Capture data. To run this program, make sure you are in the nubots home directory, then you would use: "./b nbs extract_images_MoCap ./log/<file_name>  --output ./log/<your_extraction_folder>"

Reviewer 2: This nbs tool doesn't exist? also mixture of camel case and snake case :P 
I would recommend to extract images using `./b nbs extract_images <path>` and get the numerical data via json extraction `./b nbs json <path>`.

</details>

<details>
<summary>From your computer</summary>

This method may be useful to determine if Motive is outputting packets directly or debugging depacketization without the need of a robot.

1. Ensure that the NatNet role has been installed locally.

2. Ensure that the option `message.input.MotionCapture` in DataLogging.yaml is set to true.

3. Run Natnet using `./b run natnet`

4. After your program has finished, check that a new nbs file has been created in "log/<your_program's_name>/". The filename should be the date and timestamp when the file was run.

> 5.11. You now need to extract the information you want from the nbs file. We have a few nbs decoders that have already been written for specific purposes. If one of those does not suit your needs, you will need to either modify an existing nbs extraction file, or write your own. The location of these nbs files is "nubots/tools/nbs_tools/". For example purposes, let's assume you want to extract images and Motion Capture data. To run this program, make sure you are in the nubots home directory, then you would use: "./b nbs extract_images_MoCap ./log/<file_name>  --output ./log/<your_extraction_folder>"

Reviewer 2: ```suggestion
5. You now need to extract the information you want from the nbs file. We have a few nbs decoders that have already been written for specific purposes. If one of those does not suit your needs, you will need to either modify an existing nbs extraction file, or write your own. The location of these nbs files is "nubots/tools/nbs_tools/". For example purposes, let's assume you want to extract images and Motion Capture data. To run this program, make sure you are in the nubots home directory, then you would use: "./b nbs extract_images_MoCap ./log/<file_name>  --output ./log/<your_extraction_folder>"
```
